src/services/DisposableBase.ts

```diff
@@ -0,0 +1,55 @@
+import * as vscode from 'vscode';
+import type { IDisposable } from '../types';
+import { Logger } from '../util';
+
+/** Base class for disposing of dependencies. */
+export abstract class DisposableBase implements IDisposable {
+  constructor() {
+    this._logger = new Logger(`${this.constructor.name}: DisposableBase`);
+  }
+
+  protected readonly disposables = new Set<
+    (() => void) | IDisposable | vscode.Disposable
+  >();
+  private readonly _logger: Logger;
+
+  private _isDisposed = false;
+  private _isDisposing = false;
+  private _disposalPromise: Promise<any> | undefined;
+
+  /**
+   * Dispose of dependencies. Returns immediately if already called.
+   * @returns Promise that resolves when disposal is complete.
+   */
+  async dispose(): Promise<void> {
+    if (this._isDisposing) {
+      this._logger.debug2('Already disposing');
+      return this._disposalPromise;
+    }
+
+    if (this._isDisposed) {
+      this._logger.debug2('Already disposed');
+      return this._disposalPromise;
+    }
+
+    this._logger.debug2('Disposing');
+    this._isDisposing = true;
```

Comment:
There's a race here - you set `_isDisposing` to `true` here, but then `_disposalPromise` isn't set until much later (notably after the `await this.onDisposing()`).

You should be assigning `_disposalPromise` first. This is actually a decent case for the `Promise.withResolvers`, I think. E.g.
```typescript
let { promise, resolve, reject } = Promise.withResolvers;
this._disposablePromise = promise;
...
resolve(Promise.all(disposing));
return this._disposalPromise;
```